### PR TITLE
Track cabal.project provenance for error reporting

### LINKED ISSUE #3636: new-build produces bad error message on missing cabal.project
If the cabal.project file is missing `cabal new-build` produces the rather opaque error message

```
BadPackageLocations [BadLocGlobEmptyMatch "./*.cabal"]
```

It should probably print something more informative. :)

### PR DESCRIPTION
**Pull request by brendanhay**
This changes the error when a `cabal.project` is missing from:

```
% cabal new-build
The package location glob 'foo./*.cabal' does not match any files or directories.
```

To (Implicit):

```
% cabal new-build
When using an implicit cabal.project configuration, the following errors were encountered:
The package location glob './*.cabal' does not match any files or directories.
```

Or (Explicit):

```
% cabal new-build
When using the configuration located at /home/bren/proj/cabal/cabal.project, the following errors were encountered:
The package location glob './*.cabal' does not match any files or directories.
```

This is done by tracking the provenance of the `cabal.project` file in `readProjectLocalConfig`, otherwise it defaults to `Implicit`.

The `ProjectConfig` is now also passed to `BadPackageLocations` and can be utilised in `renderBadPackageLocation`, although only the provenance is used currently to formulate a title specifying what configuration was used. This will hopefully allow individual calls to `renderBadPackageLocation` to be much more informative.

Note: I'm not certain the `Semigroup` and `Monoid` instances make sense for `ProjectConfigProvenance`.

Fixes #3636

### CODE REVIEW

**Comment by brendanhay** [comment]
I notice a couple of the `ProjectConfig` properties are failing, I'll look into it shortly.

**Comment by dcoutts** [comment]
Great!

I think we can probably improve the wording further, especially in the implicit case. In the implicit case we can match on the common kinds of errors and give a specific message.

**Comment by ezyang** [comment]
Thank you @brendanhay, provenance tracking is a great idea and will help our error messages.

You comment that you are not sure the Semigroup/Monoid instances make sense. I think this is indicative of deeper problems with the representation you have chosen. The big problem is that two `ProjectConfig`s can be merged together and now your provenance information is a bit scrambled. Provenance should not be a single `ProjectConfigProvenance`; it should be a *set*, tracking all of the `ProjectConfig`s that were inputs to the merge. Now you don't need a Semigroup/Monoid instance for `ProjectConfigProvenance`.

This also points out that your provenance tracking might be too coarse. When we `renderBadPackageLocations`, we don't care about the provenance of the config file as a whole; we just want to know which config file contributed the bad one. So really `BadPackageLocation` should be annotated with a location.

If we follow this line of reasoning to its logical conclusion, it would seem that we should refactor many of the types in Cabal to be *located*, in the same way that GHC associates every Haskell-level expression with a line and column number. That would be a big refactor, but a worthwhile one, in my opinion. But let us not let perfect be the enemy of good. If there's a bad package location, giving the user a list of config files to look at to find the offending one is better than one.

To summarize: EITHER change `ProjectConfigProvenance` into `Set ProjectConfigProvenance` and drop the bogus `Monoid`/`Semigroup` instances, OR refactor `ProjectConfig` and the data types it refers to so that everything is located. (But not the Cabal library types, because that would be BC-breaking.)

**Comment by brendanhay** [comment]
@ezyang Yes, that makes much more sense.

I interpret your comment as meaning your long-term preference is to have the location information, In which case one suggestion I have is to implement the `Set ProjectConfigProvenance` as a resolution to this PR, and then to work on adding location information as a separate PR which will then possibly supersede some of these changes.

**Comment by ezyang** [comment]
Yes. I wanted to mention the long term refactor because, as you say, it *does* supersede the easier version. But in this case the easy fix is pretty easy, so it shouldn't be too bad.

P.S., the test failure is because provenance isn't preserved if you print out a config, and then read it back in. I'm not sure if this means you should print out the provenance, or just ignore provenance when testing if it was roundtripped. Given the long term plan, probably the latter?

**Comment by brendanhay** [comment]
Both [`SavedConfig`](https://github.com/haskell/cabal/blob/8c400a40a052f466c5528fc1b9ebf32b5d1309df/cabal-install/Distribution/Client/ProjectConfig/Legacy.hs#L184) and [`LegacyProjectConfig`](https://github.com/haskell/cabal/blob/8c400a40a052f466c5528fc1b9ebf32b5d1309df/cabal-install/Distribution/Client/ProjectConfig/Legacy.hs#L223) would be represented by `mempty :: Set ProjectConfigProvenance`, but what about [`defaultImplicitProjectConfig`](https://github.com/haskell/cabal/blob/8c400a40a052f466c5528fc1b9ebf32b5d1309df/cabal-install/Distribution/Client/ProjectConfig.hs#L398)? Is it explicitly `Set.singleton Implicit`  (*drumroll*) or can that also be considered to be equivalent to an empty `Set`?

In which case, would it make more sense to use a type such as `Set FilePath` to track provenance with the empty set denoting both legacy configuration, `mempty` / default configuration, or anything excluding a configuration that was explicitly read from a `FilePath` location?

An alternative would be to track with even more granularity things such as default configuration created via `mempty` vs `defaultImplicitProjectConfig` and so on. Although it seems that way has escher-esque potential and should really be left to more general/pervasive type location annotations.

**Comment by ezyang** [comment]
In situations like this, I find it is useful to remember that `mempty` has a very specific meaning: it must be that `mappend mempty e` is `e`. So mempty MUST be the empty set.

On the other hand, it's fine for the implicit config to come with a provenance. It actually would be quite useful, because the implicit config is NOT used when there's an explicit config. So if there was an error related to implicit config, we could give an appropriate message in that case.

File: cabal-install/Distribution/Client/ProjectConfig.hs [hunk 1]
```diff
@@ -401,9 +404,16 @@ readProjectLocalConfig verbosity projectRootDir = do
...
 
         -- This is to automatically pick up deps that we unpack locally.
-        projectPackagesOptional = [ "./*/*.cabal" ]
+        projectPackagesOptional = [ "./*/*.cabal" ],
+
+        projectConfigProvenance = Set.singleton Implicit
       }
 
+    setExplicitProjectConfig :: ProjectConfig -> ProjectConfig
+    setExplicitProjectConfig config =
```
Comment (ezyang) [inline cabal-install/Distribution/Client/ProjectConfig.hs]: Are you sure you don't want to make a helper function for this operation? The helper could be more robust: instead of obliterating the old provenance it would just add to it (unless obliterating it is what you want here? Hard to tell.)

File: cabal-install/Distribution/Client/ProjectConfig.hs [hunk 2]
```diff
@@ -567,15 +578,45 @@ data BadPackageLocationMatch
...
 -- were listed, to use in error messages
---TODO: [nice to have] keep track in the ProjectConfig and BadPackageLocations
--- of whether the project config was explicit or implicit so we can report a
--- better message, either pointing to the file or talking about the issues
--- related to having no project file and no package.
+
+-- | Render bad package location error information for the implicit
+-- @cabal.project@ configuration.
+--
+-- TODO: This is currently a placeholder with the intention to supply
```
Comment (ezyang) [inline cabal-install/Distribution/Client/ProjectConfig.hs]: OK, well, let's not forget about the todo :)

File: cabal-install/Distribution/Client/ProjectConfig/Types.hs [hunk 3]
```diff
@@ -186,6 +189,18 @@ data ProjectConfigShared
   deriving (Eq, Show, Generic)
 
 
+-- | Specifies the provenance of project configuration, whether defaults were
+-- used or if the configuration was read from an explicit file path.
+data ProjectConfigProvenance
+
+     -- | The project configuration was implied.
```
Comment (ezyang) [inline cabal-install/Distribution/Client/ProjectConfig/Types.hs]: Implied by what? If I read this, I would have no idea where to look to find out more. Crossref please.

**Comment by ezyang** [comment]
@brendanhay For future reference, participants don't get emails when you push new commits to the branch, you'll have to make a comment to notice it.

**Comment by brendanhay** [comment]
Yeah, was aware my sneaky commits wouldn't be noticed. Still want to add some actual nice errors where the TODO is first. I'll incorporate the other changes you suggested too.

File: cabal-install/Distribution/Client/ProjectConfig.hs (hunk 1 above)
Comment (brendanhay) [inline cabal-install/Distribution/Client/ProjectConfig.hs]: When you say 'helper function' I'm going to assume you mean moving it to the top-level, which since it's not being used elsewhere it's nice (IMO) to have that apparent by its location in a where clause.

The old provenance should indeed be preserved via `Set.insert`.

File: cabal-install/Distribution/Client/ProjectConfig.hs [hunk 4]
```diff
@@ -614,9 +614,14 @@ renderBadPackageLocations (BadPackageLocations prov bpls)
...
-    case bpl of
-        _ -> renderBadPackageLocation bpl
+renderImplicitBadPackageLocation bpl = case bpl of
+    BadLocGlobEmptyMatch pkglocstr ->
+        "No cabal.project or cabal file matching the glob '"
+     ++ pkglocstr ++ "' was found.\n"
+     ++ "Please create a package description file <pkgname>.cabal "
+     ++ "or a cabal.project file referencing the packages you  "
+     ++ "want to build."
+    _ -> renderBadPackageLocation bpl
```
Comment (brendanhay) [inline cabal-install/Distribution/Client/ProjectConfig.hs]:
```
No cabal.project or cabal file matching the glob './*.cabal' was found.
Please create a package description file <pkgname>.cabal or a cabal.project
file referencing the packages you want to build.
```

Comment (brendanhay) [inline cabal-install/Distribution/Client/ProjectConfig.hs]: Regarding the other cases, I'm not convinced I actually understand well enough under what situations these errors are raised to correctly word the implicit error message.

File: cabal-install/Distribution/Client/ProjectConfig.hs [hunk 5]
```diff
@@ -567,15 +578,50 @@ data BadPackageLocationMatch
...
+      -- using possibly specialized error messages.
+    | provenance == [Implicit] = renderErrors renderImplicitBadPackageLocation
+
+      -- The configuration was read from one or more explicit path(s),
+      -- list the locations and render the bad package error information.
+      -- The intent is to supersede this with the relevant location information
+      -- per package error.
+    | otherwise =
+           "When using configuration(s) from "
+        ++ intercalate ", " (mapMaybe getExplicit provenance)
```
Comment (ezyang) [inline cabal-install/Distribution/Client/ProjectConfig.hs]: Maybe this should also mention if the implicit is used? (I think this case is impossible, so outputting something in that case could help debugging in the future. Or maybe even an error.)

File: cabal-install/Distribution/Client/ProjectConfig.hs [hunk 6]
```diff
@@ -567,15 +578,50 @@ data BadPackageLocationMatch
...
+-- | Render bad package location error information for the implicit
+-- @cabal.project@ configuration.
+--
+-- TODO: This is currently a placeholder with the intention to supply
+-- informative help text about the issues related to having no project
+-- file and no package.
+renderImplicitBadPackageLocation :: BadPackageLocation -> String
+renderImplicitBadPackageLocation bpl = case bpl of
+    BadLocGlobEmptyMatch pkglocstr ->
+        "No cabal.project or cabal file matching the glob '"
```
Comment (ezyang) [inline cabal-install/Distribution/Client/ProjectConfig.hs]: Perhaps say "default glob" rather than "glob"? Because it doesn't actually exist anywhere.

**Comment by ezyang** [comment]
Still looking good :)
